dialects/scf/src/canonicalization/fold_redundant_yields.rs

```diff
@@ -0,0 +1,966 @@
+use alloc::rc::Rc;
+use core::{any::TypeId, ops::Index};
+
+use midenc_hir::{
+    patterns::{Pattern, PatternBenefit, PatternInfo, PatternKind, RewritePattern},
+    *,
+};
+
+use crate::*;
+
+/// Lift common yield results from their regions to their predecessors.
+pub struct FoldRedundantYields {
+    info: PatternInfo,
+}
+
+impl FoldRedundantYields {
+    pub fn new(context: Rc<Context>) -> Self {
+        Self {
+            info: PatternInfo::new(
+                context,
+                "fold-redundant-yields",
+                PatternKind::Trait(TypeId::of::<dyn RegionBranchOpInterface>()),
+                PatternBenefit::MAX,
+            ),
+        }
+    }
+}
+
+impl Pattern for FoldRedundantYields {
+    fn info(&self) -> &PatternInfo {
+        &self.info
+    }
+}
+
+impl RewritePattern for FoldRedundantYields {
+    fn match_and_rewrite(
+        &self,
+        mut op_ref: OperationRef,
+        rewriter: &mut dyn Rewriter,
+    ) -> Result<bool, Report> {
+        let mut op = op_ref.borrow_mut();
+
+        let Some(br_op) = op.as_trait::<dyn RegionBranchOpInterface>() else {
+            return Ok(false);
+        };
+
+        // For each successor gather its terminator and the terminator operands.
+        let mut term_ops: SmallVec<[_; 4]> = SmallVec::new();
+        let mut region_yields: SmallVec<[_; 4]> = SmallVec::new();
+
+        for succ_region in br_op.get_successor_regions(RegionBranchPoint::Parent) {
+            let Some(region_ref) = succ_region.successor() else {
+                return Ok(false);
+            };
+
+            // Several assertions follow: an SCF region always has a single block, that block must
+            // have a terminator which then implements RegionBranchTerminatorOpInterface.
+            let region = region_ref.borrow();
+            assert!(region.has_one_block());
+
+            let block = region.entry();
+            let term_op_ref =
+                block.terminator().expect("All region blocks must have a terminator.");
+
+            // Got the terminator.
+            let term_op = term_op_ref.borrow();
+            let term_op = term_op.as_trait::<dyn RegionBranchTerminatorOpInterface>().expect(
+                "All region block terminators must impl RegionBranchTerminatorOpInterface.",
+            );
+
+            // For now we only support regions with a simple `yield` terminator.  This may change
+            // in the future if/when we support other RegionBranchOps (e.g., `while`).
+            let term_op_name = term_op.as_operation().name();
+            if term_op_name.dialect() != "scf" || term_op_name.name() != "yield" {
+                return Ok(false);
+            }
```

Comment:
```suggestion
            if !term_op.as_operation().is::<crate::YieldOp>() {
                return Ok(false);
            }
```

Much easier to use `Operation::is::<T>()` for this sort of thing 🙂

Reply:
Shit, sorry, forgot the op name is `Yield`, not `YieldOp` 😬